system/lib/libc/musl/src/thread/pthread_join.c

```diff
@@ -7,26 +7,59 @@ static void dummy1(pthread_t t)
 }
 weak_alias(dummy1, __tl_sync);
 
+#ifdef __EMSCRIPTEN__ // XXX Emscripten add extern for __emscripten_thread_cleanup
+extern void __emscripten_thread_cleanup(pthread_t thread);
+#endif
+
 static int __pthread_timedjoin_np(pthread_t t, void **res, const struct timespec *at)
 {
+#ifdef __EMSCRIPTEN__
+	// Attempt to join a thread which does not point to a valid thread, or
+	// does not exist anymore.
+	if (t->self != t) return ESRCH;
+	// Thread is attempting to join to itself.  Already detached threads are
+	// handled below by returning EINVAL instead.
```

Comment:
Maybe mention that these extra checks are because we want to pass the posixtest suite?   IIRC both these cases count as undefined behaviour so are not strictly necessary.

Reply:
Do you mean the extra `t->detach_state != DT_DETACHED` check? AFAIK, that was needed for our `test_main_pthread_join_detach` and `test_proxy_pthread_join_detach` tests, and not needed for the posix tests.

I could also remove this check and update that test exception by partially reverting commit ca90a4b17f8a483474b8e6df2a441febf0413f88, if you want.

Reply:
Lets leave that as a possible followup.

Reply:
I can confirm that the `t->detach_state != DT_DETACHED` check is only needed for the `other.test_{proxy,main}_pthread_join_detach` tests, see for e.g. work-in-progress commit 493f2e510ba37e6c95725ded9bc0db9637fe90ef that updates its test asserts.

Reply:
I added a TODO with commit 5e65f91c1668cc0bb43598cad2a0717fac5eab1c, since removing this check would also mean that it causes an `EDEADLK` error when already-detached threads are attempting to detach itself (since `pthread_detach` re-uses code from `pthread_join`).